From: aparna.salian <aparna.salian@globaledgesoft.com>
Subject: [PATCH] lib/hostap: Workaround for 
PROBE_REQ/ASSOC_REQ/REASSOC_REQ events

(requires a patched hostapd, 991-nl_cmd_frame-event-gen.patch)

--- a/core/src/lib/hostap/inc/opensync-hapd.h
+++ b/core/src/lib/hostap/inc/opensync-hapd.h
@@ -53,6 +53,10 @@ struct hapd {
     void (*wps_timeout)(struct hapd *hapd);
     void (*wps_disable)(struct hapd *hapd);
     void (*wpa_key_mismatch)(struct hapd *hapd, const char *mac);
+    void (*cmd_frame)(struct hapd *hapd, const char *event);
+    void (*cmd_frame_probe_req)(struct hapd *hapd, const char *event);
+    void (*cmd_frame_action)(struct hapd *hapd, const char *event);
+    void (*cmd_frame_disconnect)(struct hapd *hapd, const char *event);
     struct ctrl ctrl;
 };
 
--- a/core/src/lib/hostap/src/hapd.c
+++ b/core/src/lib/hostap/src/hapd.c
@@ -121,6 +121,7 @@ hapd_ctrl_cb(struct ctrl *ctrl, int leve
     const char *mac = NULL;
     const char *k;
     const char *v;
+    const char *type = "";
     char *args = strdupa(buf);
     char *kv;
 
@@ -132,6 +133,30 @@ hapd_ctrl_cb(struct ctrl *ctrl, int leve
      */
     event = strsep(&args, " ") ?: "_nope_";
 
+    if (!strcmp(event, EV(NL_CMD_FRAME))) {
+        while ((kv = strsep(&args, " "))) {
+            if ((k = strsep(&kv, "=")) && (v = strsep(&kv, ""))) {
+                if (!strcmp(k, "type")) {
+                    type = v;
+                    break;
+                }
+            }
+        }
+        if (!strcmp(type, EVT_FRAME_PROBE_REQ)) {
+            if (hapd->cmd_frame_probe_req)
+                hapd->cmd_frame_probe_req(hapd, args);
+        } else if (!strcmp(type, EVT_FRAME_ACTION)) {
+            if (hapd->cmd_frame_action)
+                hapd->cmd_frame_action(hapd, args);
+        } else if (!strcmp(type, EVT_FRAME_DISCONNECT)) {
+            if (hapd->cmd_frame_disconnect)
+                hapd->cmd_frame_disconnect(hapd, args);
+        } else {
+            if (hapd->cmd_frame)
+                hapd->cmd_frame(hapd, buf);
+        }
+    }
+
     if (!strcmp(event, EV(AP_STA_CONNECTED))) {
         mac = strsep(&args, " ") ?: "";
 
